Time.hs

```diff
@@ -299,35 +324,69 @@ insertIntMapStrict n0 = go n0 mempty
     go 0 acc = acc
     go n !acc = go (n - 1) (Data.IntMap.Strict.insert n n acc)
 
-insertHashTableIOBasic :: Int -> IO (Data.HashTable.IO.BasicHashTable Int Int)
-insertHashTableIOBasic n0 = do
-  ht <- Data.HashTable.IO.new
+insertHashTableIO :: Data.HashTable.Class.HashTable ht
+  => Data.HashTable.IO.IOHashTable ht Int Int
+  -> Int
+  -> IO (Data.HashTable.IO.IOHashTable ht Int Int)
+insertHashTableIO ht n0 = do
   mapM_ (\n -> Data.HashTable.IO.insert ht n n) [1..n0]
   return ht
 
-insertHashTableIOCuckoo :: Int -> IO (Data.HashTable.IO.CuckooHashTable Int Int)
-insertHashTableIOCuckoo n0 = do
-  ht <- Data.HashTable.IO.new
-  mapM_ (\n -> Data.HashTable.IO.insert ht n n) [1..n0]
-  return ht
+insertHashTableIOBasic :: Data.HashTable.IO.BasicHashTable Int Int
+  -> Int
+  -> IO (Data.HashTable.IO.BasicHashTable Int Int)
+insertHashTableIOBasic = insertHashTableIO
 
-insertHashTableIOLinear :: Int -> IO (Data.HashTable.IO.LinearHashTable Int Int)
-insertHashTableIOLinear n0 = do
-  ht <- Data.HashTable.IO.new
-  mapM_ (\n -> Data.HashTable.IO.insert ht n n) [1 .. n0]
-  return ht
+insertHashTableIOCuckoo :: Data.HashTable.IO.CuckooHashTable Int Int
+  -> Int
+  -> IO (Data.HashTable.IO.CuckooHashTable Int Int)
+insertHashTableIOCuckoo = insertHashTableIO
 
-insertJudy :: Int -> IO (Data.Judy.JudyL Int)
-insertJudy n0 = do
-  j <- Data.Judy.new
+insertHashTableIOLinear :: Data.HashTable.IO.LinearHashTable Int Int
+  -> Int
+  -> IO (Data.HashTable.IO.LinearHashTable Int Int)
+insertHashTableIOLinear = insertHashTableIO
+
+insertJudy :: Data.Judy.JudyL Int -> Int -> IO (Data.Judy.JudyL Int)
+insertJudy j n0 = do
   mapM_ (\n -> Data.Judy.insert (fromIntegral n) n j) [1 .. n0]
   return j
 
-judyFromList :: [(Int,Int)] -> IO (Data.Judy.JudyL Int)
-judyFromList xs = do
+judyLookup :: Data.Judy.JudyL Int -> Int -> IO (Maybe Int)
+judyLookup j k = Data.Judy.lookup (fromIntegral k) j
+
+intersectionJudy :: Data.Judy.JudyL Int -> Data.Judy.JudyL Int -> IO (Data.Judy.JudyL Int)
+intersectionJudy ij0 ij1 = do
   j <- Data.Judy.new
-  mapM_ (\(k,v) -> Data.Judy.insert (fromIntegral k) v j) xs
+  j0 <- Data.Judy.unsafeFreeze ij0
```

Comment:
I don't think an unsafe freeze is quite a fair comparison, it should be a pure freeze. Or else maybe there should be safe/unsafe entries in the benchmarks.